File: 3510. minimum-pair-removal-to-sort-array-ii.py
from typing import List
from itertools import pairwise
import heapq


class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        n=len(nums)
        # check if sorted
        if all(x<=y for x, y in pairwise(nums)):
            return 0
        rmv=[False]*n
        prv=[i-1 for i in range(n)]
        nxt=[i+1 if i+1<n else -1 for i in range(n)]

        heap=[(nums[i]+nums[i+1], i) for i in range(n-1)]
        heapq.heapify(heap)

        bad=sum( nums[i]>nums[i+1] for i in range(n-1))
        op=0
        while bad>0:
            Sum, i=heapq.heappop(heap)

            if rmv[i] or nxt[i]==-1: continue
            j=nxt[i]
            if rmv[j] or nums[i]+nums[j]!=Sum: continue

            pi=prv[i]
            nj=nxt[j]

            # remove old violations
            if pi!=-1 and nums[pi]>nums[i]: bad-=1
            if nums[i]>nums[j]: bad-=1
            if nj!=-1 and nums[j]>nums[nj]: bad-=1

            # merge
            nums[i]=Sum
            rmv[j]=True

            nxt[i]=nj
            if nj!=-1: prv[nj]=i
            
            # add new violations
            if pi!=-1 and nums[pi]>nums[i]: bad+=1
            if nj!=-1 and nums[i]>nums[nj]: bad+=1

            # update heap
            if pi!= -1:
                heapq.heappush(heap, (nums[pi]+nums[i], pi))
            if nj!= -1:
                heapq.heappush(heap, (nums[i]+nums[nj], i))
                
            op+=1
        return op
        

# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Test case 1
    nums1 = [5,2,3,1]
    output1 = solution.minimumPairRemoval(nums1)
    print(f"Test 1: nums = [5,2,3,1]")
    print(f"Output: {output1}, Expected: 2")
    
    # Test case 2
    nums2 = [1,2,2]
    output2 = solution.minimumPairRemoval(nums2)
    print(f"Test 2: nums = [1,2,2]")
    print(f"Output: {output2}, Expected: 0")
    
    print("\nTest cases executed!")

File: 3510. minimum-pair-removal-to-sort-array-ii.java
import java.util.*;

class Solution {
    public int minimumPairRemoval(List<Integer> nums) {
        int n = nums.size();
        
        // Check if already sorted
        boolean sorted = true;
        for (int i = 0; i < n - 1; i++) {
            if (nums.get(i) > nums.get(i + 1)) {
                sorted = false;
                break;
            }
        }
        if (sorted) return 0;
        
        // Create mutable array
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = nums.get(i);
        }
        
        boolean[] rmv = new boolean[n];
        int[] prv = new int[n];
        int[] nxt = new int[n];
        
        for (int i = 0; i < n; i++) {
            prv[i] = i - 1;
            nxt[i] = (i + 1 < n) ? i + 1 : -1;
        }
        
        // Min heap: (sum, index)
        PriorityQueue<long[]> heap = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        for (int i = 0; i < n - 1; i++) {
            heap.offer(new long[]{(long)arr[i] + arr[i + 1], i});
        }
        
        int bad = 0;
        for (int i = 0; i < n - 1; i++) {
            if (arr[i] > arr[i + 1]) bad++;
        }
        
        int op = 0;
        while (bad > 0) {
            long[] entry = heap.poll();
            long Sum = entry[0];
            int i = (int)entry[1];
            
            if (rmv[i] || nxt[i] == -1) continue;
            int j = nxt[i];
            if (rmv[j] || arr[i] + arr[j] != Sum) continue;
            
            int pi = prv[i];
            int nj = nxt[j];
            
            // Remove old violations
            if (pi != -1 && arr[pi] > arr[i]) bad--;
            if (arr[i] > arr[j]) bad--;
            if (nj != -1 && arr[j] > arr[nj]) bad--;
            
            // Merge
            arr[i] = (int)Sum;
            rmv[j] = true;
            
            nxt[i] = nj;
            if (nj != -1) prv[nj] = i;
            
            // Add new violations
            if (pi != -1 && arr[pi] > arr[i]) bad++;
            if (nj != -1 && arr[i] > arr[nj]) bad++;
            
            // Update heap
            if (pi != -1) {
                heap.offer(new long[]{(long)arr[pi] + arr[i], pi});
            }
            if (nj != -1) {
                heap.offer(new long[]{(long)arr[i] + arr[nj], i});
            }
            
            op++;
        }
        return op;
    }

    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        List<Integer> nums1 = Arrays.asList(5, 2, 3, 1);
        int output1 = solution.minimumPairRemoval(nums1);
        System.out.println("Test 1: nums = [5,2,3,1]");
        System.out.println("Output: " + output1 + ", Expected: 2");
        
        // Test case 2
        List<Integer> nums2 = Arrays.asList(1, 2, 2);
        int output2 = solution.minimumPairRemoval(nums2);
        System.out.println("Test 2: nums = [1,2,2]");
        System.out.println("Output: " + output2 + ", Expected: 0");
        
        System.out.println("\nTest cases executed!");
    }
}
